File: frontend/src/types/subject.ts
import type { UUID } from './common';

export interface Subject {
  id: UUID;
  subject_number: string;
  scenario_id: UUID | null;
  cohort_id: string | null;
  arm_id: string | null;
  site_node_id: UUID | null;
  status: string;
  screened_at: string | null;
  enrolled_at: string | null;
  discontinued_at: string | null;
  completed_at: string | null;
  notes: string | null;
  attributes: Record<string, unknown> | null;
  created_at: string;
}

export interface SubjectCreate {
  subject_number: string;
  scenario_id?: UUID;
  cohort_id?: string;
  arm_id?: string;
  site_node_id?: string;
  status?: string;
  screened_at?: string;
  notes?: string;
  attributes?: Record<string, unknown>;
}

export interface SubjectVisit {
  id: UUID;
  subject_id: UUID;
  visit_id: string;
  scheduled_date: string | null;
  actual_date: string | null;
  status: string;
  notes: string | null;
}

export interface SubjectVisitCreate {
  visit_id: string;
  scheduled_date?: string;
  status?: string;
  notes?: string;
}

export interface KitAssignment {
  id: UUID;
  subject_visit_id: UUID;
  lot_id: UUID | null;
  product_id: string;
  presentation_id: string | null;
  qty_dispensed: number;
  dispensed_at: string | null;
  dispensed_by: string | null;
  returned_qty: number;
  returned_at: string | null;
}

export interface DispenseRequest {
  lot_id?: UUID;
  product_id: string;
  presentation_id?: string;
  qty_dispensed: number;
  dispensed_by?: string;
}

export interface KitReturnRequest {
  returned_qty: number;
}

export type SubjectStatus = 'SCREENED' | 'ENROLLED' | 'ACTIVE' | 'DISCONTINUED' | 'COMPLETED';
export type VisitStatus = 'SCHEDULED' | 'COMPLETED' | 'MISSED' | 'CANCELLED';

File: app/models/subject.py
import uuid
from datetime import UTC, datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Subject(Base):
    """A trial subject enrolled at a site."""

    __tablename__ = "subjects"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    subject_number: Mapped[str] = mapped_column(String(64), unique=True, index=True)
    scenario_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("scenarios.id"), nullable=True, index=True
    )
    cohort_id: Mapped[str | None] = mapped_column(String(64), nullable=True, index=True)
    arm_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    site_node_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_nodes.id"), nullable=True
    )

    status: Mapped[str] = mapped_column(String(32), default="SCREENED", index=True)
    # SCREENED → ENROLLED → ACTIVE → DISCONTINUED | COMPLETED

    screened_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    enrolled_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    discontinued_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    attributes: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(UTC))

    visits: Mapped[list["SubjectVisit"]] = relationship(back_populates="subject", cascade="all, delete-orphan")


class SubjectVisit(Base):
    """A scheduled or completed visit for a subject."""

    __tablename__ = "subject_visits"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    subject_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("subjects.id"), index=True)

    visit_id: Mapped[str] = mapped_column(String(64))  # e.g. "C1D1", "MAINT_QW_w5"
    scheduled_date: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    actual_date: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    status: Mapped[str] = mapped_column(String(32), default="SCHEDULED")
    # SCHEDULED → COMPLETED | MISSED | CANCELLED

    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    subject: Mapped["Subject"] = relationship(back_populates="visits")
    kit_assignments: Mapped[list["KitAssignment"]] = relationship(
        back_populates="subject_visit", cascade="all, delete-orphan"
    )


class KitAssignment(Base):
    """A record of product dispensed (or returned) at a visit."""

    __tablename__ = "kit_assignments"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    subject_visit_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("subject_visits.id"), index=True
    )
    lot_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("inventory_lots.id"), nullable=True
    )

    product_id: Mapped[str] = mapped_column(String(128))
    presentation_id: Mapped[str | None] = mapped_column(String(128), nullable=True)
    qty_dispensed: Mapped[float] = mapped_column(Float, default=0.0)

    dispensed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    dispensed_by: Mapped[str | None] = mapped_column(String(128), nullable=True)

    returned_qty: Mapped[float] = mapped_column(Float, default=0.0)
    returned_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    subject_visit: Mapped["SubjectVisit"] = relationship(back_populates="kit_assignments")
